gnovm/pkg/doc/dirs.go

```diff
@@ -0,0 +1,163 @@
+// only walks the tree once, caching the data it finds.
+type Dirs struct {
+	scan   chan Dir // Directories generated by walk.
+	hist   []Dir    // History of reported Dirs.
+	offset int      // Counter for Next.
+}
+
+// NewDirs begins scanning the given stdlibs directory.
+func NewDirs(dirs ...string) *Dirs {
+	d := &Dirs{
+		hist: make([]Dir, 0, 256),
+		scan: make(chan Dir),
+	}
+	go d.walk(dirs)
+	return d
+}
+
+// Reset puts the scan back at the beginning.
+func (d *Dirs) Reset() {
+	d.offset = 0
+}
+
+// Next returns the next directory in the scan. The boolean
+// is false when the scan is done.
+func (d *Dirs) Next() (Dir, bool) {
+	if d.offset < len(d.hist) {
+		dir := d.hist[d.offset]
+		d.offset++
+		return dir, true
+	}
+	dir, ok := <-d.scan
+	if !ok {
+		return Dir{}, false
+	}
+	d.hist = append(d.hist, dir)
+	d.offset++
+	return dir, ok
+}
+
+// walk walks the trees in the given roots.
+func (d *Dirs) walk(roots []string) {
+	for _, root := range roots {
+		d.bfsWalkRoot(root)
+	}
+	close(d.scan)
+}
+
+// bfsWalkRoot walks a single directory hierarchy in breadth-first lexical order.
+// Each Go source directory it finds is delivered on d.scan.
+func (d *Dirs) bfsWalkRoot(root string) {
+	root = filepath.Clean(root)
+
+	// this is the queue of directories to examine in this pass.
+	this := []string{}
+	// next is the queue of directories to examine in the next pass.
+	next := []string{root}
+
+	for len(next) > 0 {
+		this, next = next, this[:0]
+		for _, dir := range this {
+			fd, err := os.Open(dir)
+			if err != nil {
+				log.Print(err)
+				continue
+			}
+			entries, err := fd.Readdir(0)
+			fd.Close()
+			if err != nil {
+				log.Print(err)
+				continue
+			}
+			hasGnoFiles := false
+			for _, entry := range entries {
+				name := entry.Name()
+				// For plain files, remember if this directory contains any .gno
+				// source files, but ignore them otherwise.
+				if !entry.IsDir() {
+					if !hasGnoFiles && strings.HasSuffix(name, ".gno") {
+						hasGnoFiles = true
+					}
+					continue
+				}
+				// Entry is a directory.
+
+				// Ignore same directories ignored by the go tool.
+				if name[0] == '.' || name[0] == '_' || name == "testdata" {
+					continue
+				}
+				// Remember this (fully qualified) directory for the next pass.
+				next = append(next, filepath.Join(dir, name))
+			}
+			if hasGnoFiles {
+				// It's a candidate.
+				var importPath string
+				if len(dir) > len(root) {
+					importPath = filepath.ToSlash(dir[len(root)+1:])
+				}
+				d.scan <- Dir{importPath, dir}
+			}
+		}
+	}
+}
+
+// findPackage finds a package iterating over d where the import path has
+// name as a suffix (which may be a package name or a fully-qualified path).
+// returns a list of possible directories. If a directory's import path matched
+// exactly, it will be returned as first.
+func (d *Dirs) findPackage(name string) []Dir {
+	d.Reset()
+	candidates := make([]Dir, 0, 4)
```

Comment:
why 4?

Reply:
Same logic as the other one, here it's because while the candidates will generally be just 1, it could be a few